runtime/plaid/src/apis/aws/dynamodb.rs

```diff
@@ -0,0 +1,745 @@
+    sync::Arc,
+};
+
+use serde::{Deserialize, Serialize};
+
+use crate::apis::ApiError;
+use crate::{get_aws_sdk_config, loader::PlaidModule, AwsAuthentication};
+
+/// Defines configuration for the DynamoDB API
+#[derive(Deserialize, Serialize)]
+pub struct DynamoDbConfig {
+    /// Enable dynamodb local endpoint for testing
+    local_endpoint: bool,
+    /// This can either be:
+    /// - `IAM`: Uses the IAM role assigned to the instance or environment.
+    /// - `ApiKey`: Uses explicit credentials, including an access key ID, secret access key, and region.
+    #[serde(skip_serializing)]
+    authentication: AwsAuthentication,
+    /// Configured writers - maps a table name to a list of rules that are allowed to READ or WRITE data
+    rw: HashMap<String, HashSet<String>>,
+    /// Configured readers - maps a table name to a list of rules that are allowed to READ data
+    r: HashMap<String, HashSet<String>>,
+}
+
+/// Represents the DynamoDB API client.
+/// NOTE: if Plaid is configured with the DynamoDB database backend, sharing tables here will lead to undefined behaviour
+pub struct DynamoDb {
+    /// The underlying client used to interact with the KMS API.
+    client: Client,
+    /// Configured writers - maps a table name to a list of rules that are allowed to READ or WRITE data
+    rw: HashMap<String, HashSet<String>>,
+    /// Configured readers - maps a table name to a list of rules that are allowed to READ data
+    r: HashMap<String, HashSet<String>>,
+}
+
+#[derive(PartialEq, PartialOrd)]
+enum AccessScope {
+    Read,
+    Write,
+}
+
+impl DynamoDb {
+    /// Creates a new instance of `DynamoDb`
+    pub async fn new(config: DynamoDbConfig) -> Self {
+        let DynamoDbConfig {
+            authentication,
+            rw,
+            r,
+            local_endpoint,
+        } = config;
+
+        if local_endpoint {
+            return DynamoDb::local_endpoint(r, rw).await;
+        }
+
+        let sdk_config = get_aws_sdk_config(&authentication).await;
+        let client = Client::new(&sdk_config);
+
+        Self { client, rw, r }
+    }
+
+    /// constructor for the local instance of DynamoDB
+    async fn local_endpoint(
+        r: HashMap<String, HashSet<String>>,
+        rw: HashMap<String, HashSet<String>>,
+    ) -> Self {
+        let config = aws_config::defaults(aws_config::BehaviorVersion::latest())
+            // DynamoDB run locally uses port 8000 by default.
+            .endpoint_url("http://localhost:8000")
+            .load()
+            .await;
+        let dynamodb_local_config = aws_sdk_dynamodb::config::Builder::from(&config).build();
+
+        let client = aws_sdk_dynamodb::Client::from_conf(dynamodb_local_config);
+
+        Self { client, r, rw }
+    }
+
+    fn check_module_permissions(
+        &self,
+        access_scope: AccessScope,
+        module: Arc<PlaidModule>,
+        table_name: &str,
+    ) -> Result<(), ApiError> {
+        match access_scope {
```

Comment:
Design musing: You could check both and then have a match statement that covers all cases (including the missing one attempts to write but only has read).

That would allow you to emit a `warn!` which would be slightly more clear that "doesn't have permission" in "doesn't have write, only has read".